### PR TITLE
[RFC] include/interval_set: iterate until p->first < start

### PR DESCRIPTION
**Pull request by dparmar18**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by dparmar18** [comment]
there are a couple of crash trackers where assertion fails at `ceph_assert(p->first <= start);` in `void erase()`. The places are different but the assertion failure is same. Failures at these places will definitely be addressed but I feel these helpers need some changes too. 

Existing code does decrement by one in case the key is gt than `start` but only once and so the decremented element's key should still be gt than `start`. I.e. the single decrement might not be enough, therefore keep decrementing until `p->first` is less than `start`; break and return.

https://tracker.ceph.com/issues/62356
https://tracker.ceph.com/issues/54943
https://tracker.ceph.com/issues/60241

**Comment by dparmar18** [comment]
@lxbsz

File: src/include/interval_set.h [hunk 1]
```diff
@@ -249,44 +249,60 @@ class interval_set {
  private:
   auto find_inc(T start) const {
     auto p = m.lower_bound(start);  // p->first >= start
-    if (p != m.begin() &&
-        (p == m.end() || p->first > start)) {
-      --p;   // might overlap?
-      if (p->first + p->second <= start)
-        ++p; // it doesn't.
+    while (p != m.begin()) {
+        if (p->first <= start) {
```
Comment (lxbsz) [inline src/include/interval_set.h:253]: This is buggy. Old code will make sure that the `p` won't be `.end()`, but yours seems will crash ?

Comment (dparmar18) [reply src/include/interval_set.h:253]: so if we dont hit this `if {}` then it assume either `p->first > start` or `p==m.end()`, so in this case we decrement `--p`

Comment (dparmar18) [reply src/include/interval_set.h:253]: i.e. it wont hit `end()` here either

Comment (lxbsz) [reply src/include/interval_set.h:253]: Won't the `m.lower_bound()` return `.end()` if it couldn't find any item ?

Comment (lxbsz) [reply src/include/interval_set.h:253]: Yeah, the corner edge case.

File: src/include/interval_set.h [hunk 2]
```diff
@@ -249,44 +249,60 @@ class interval_set {
...
+        }
+    }
+    if (p->first + p->second <= start) {
+      ++p; // it doesn't.
     }
     return p;
   }
   
   auto find_inc_m(T start) {
     auto p = m.lower_bound(start);
```
Comment (lxbsz) [inline src/include/interval_set.h:266]: I am not sure why the existing code use `lower_bound()`, IMO we should switch to use the `upper_bound()` instead.

For example:

```
[40] = 10
[60] = 10
[100] = 30
```
If the `start` is `40` the `lower_bound()` will return `.end()`, right ? Then the `--p` will be `[100] = 30` ?

While the `upper_bound()` will return `[40] = 10`.

File: src/include/interval_set.h (hunk 1 above)
Comment (dparmar18) [reply src/include/interval_set.h:254]: check now

File: src/include/interval_set.h (hunk 2 above)
Comment (lxbsz) [reply src/include/interval_set.h:268]: Okay, you are right. I misread it. Thanks.

File: src/include/interval_set.h (hunk 1 above)
Comment (lxbsz) [reply src/include/interval_set.h:254]: In `erase()` it will check `ceph_assert(p != m.end()); `. Your above change could return `.end()`.

**Comment by dparmar18** [comment]
This PR doesn't make sense now, the working of `lower_bound` is such that it will return an iterator pointing to an element that is just greater than the `start`, so the `p` doesn't need to be iterated through the entire map; just a single decrement is enough.

This doesn't mean the interval_set code is entirely perfect, the `lower_bound()` code has to do the adjustments `--p` / `++p` to make sure that element the iterator is pointing to is correct. The code definitely needs some corrections for sure but since the priority is to send a fix for the mentioned crash trackers, i'll leave this for some other day. Closing this PR for now, if anyone has any comments/suggestions to add; please feel free to do it anytime.

**Pull request closed without merging**